### ISSUE TITLE
Should we allow accesses via unchecked pointers to struct members that are checked arrays?

### ISSUE THREAD

**Issue by dtarditi** [checkedc/checkedc#197]
I'm interested in hearing people's thoughts about the following issue.

A programmer can declare a struct S with a checked array member:
```
struct S {
  int arr checked[10];
}
```
This array can be accessed using an unchecked pointer:
```
int f(struct S *p) {
  p->arr[5];
}
```
The bounds check will guarantee that the memory access is within `p->arr`.   However, when p is an unchecked pointer, there is no guarantee that `p->arr` is valid memory to access.    

There is another way of explaining the issue.  Given `e->arr` (where `e`  is non-modifying), the bounds are `bounds(e->arr, e->arr + 1)`.   However, if `e` is an unchecked pointer, this means that the we now have a use of an unchecked pointer in the bounds.  Bounds are always supposed to be trusted. If we don't trust unchecked pointers variables to be in range, should we trust bounds that depend on the values of those variables?

Three possible approaches are:
1. Allow this and don't warn.   For an access e->arr, the rule is that we first check that e is a range.   We then produce bounds that depends on e.  If e has unchecked pointer type, it is understood that e might not be a range.    This allows programmers to easily enable bounds checks for accesses to arrays embedded in structs, even if the pointers to those structs may not be valid.
2. Allow this but suggest that a compiler warn about these kinds of accesses in unchecked code.
3. Forbid this: we shouldn't allow expressions to produce bounds that depend on unchecked pointers being in range.

Note that pointers to checked arrays have a similar issue:
```
int (*p) checked[10]
```
The expression `*p` will have the type `int checked[10]`.   This will be converted implicitly to an `array_ptr<int>` that points to an array of 10 elements.  We end up with bounds that depend on an unchecked pointer type.  The bounds are `bounds(*p, *p + 10)`.

**Comment by dtarditi** [comment]
Correct title.

**Comment by dtarditi** [comment]
I discussed this with @awruef and @mwhicks1 on Thursday.   Mike suggested that we disallow unchecked pointers to checked arrays, but allow unchecked pointers to structs containing checked arrays.

**Comment by dtarditi** [comment]
@bharadwajy and I discussed this the other week and I thought about it some more.   There's another possible solution, which is to change the typing for the dereference operator applied to a pointer to an array.

It is useful to do a case analysis of the possible combinations and their meanings:
1. Checked pointer to checked array: there are no problems with this case.  Checking is preserved.
2. Unchecked pointer to checked array.   The dereference operator will produce a checked array.  This will then be converted to a checked pointer.   The bounds should be the bounds for the original unchecked pointer.    So we'll be able to access memory only if we know bounds for the original unchecked pointer, which is reasonable.  We don't lose checking.
3. Checked pointer to unchecked array.  Under current rules, the dereference operator will produce an unchecked array.  This is problematic because it means that checking has been lost.   We could change the typing rules to say that the dereference operator produces a checked array type in this case.  Then everything else will work out.
4. Unchecked pointer to unchecked array: there are no problems in this case.  This is the original unchecked case.

By this analysis, there's actually nothing wrong with unchecked pointers to checked arrays.   By changing the typing rules, we can avoid checks being discarded.

That still leaves us to ponder the meaning of an unchecked pointer to a struct member that is an array.  It has bounds, but the bounds are not to trusted.

**Comment by dtarditi** [comment]
Disallowing unchecked pointers to checked arrays would break the incremental conversion story. Programmers who want guarantees should place code in checked scopes, which disallow unchecked pointers.
